Win32-network/src/System/Win32/Async/IOManager.hsc

```diff
@@ -0,0 +1,284 @@
+      -- 'dequeueCompletionPackets' is ment to run for the entire execution time
+      -- of an appliction.
+      GQCSResult { gqcsResult
+                 , gqcsOverlappedIsNull
+                 , gqcsCompletionKey
+                 , gqcsIODataPtr
+                 }
+          <- getQueuedCompletionStatus port numBytesPtr maxBound
+      errorCode <- Win32.getLastError
+
+      -- gqcsIODataPtr was allocated by 'withIODataPtr' or
+      -- 'wsaWaitForCompletion', and we are responsible to deallocate it but
+      -- only if 'gqcsCompletionKey' is 'True'.
+
+      if | gqcsOverlappedIsNull && errorCode == eRROR_ABANDONED_WAIT_0 ->
+           -- this is triggered by closing IO completion 'HANDLE' while there
+           -- are outstanding requests (this is documented
+           -- <https://docs.microsoft.com/en-us/windows/win32/api/ioapiset/nf-ioapiset-getqueuedcompletionstatus#remarks>)
+              return ()
+         | gqcsOverlappedIsNull && errorCode == eRROR_INVALID_HANDLE ->
+           -- If one doesn't terminate the dequeueing thread, some of the tests
+           -- will take 10x-100x more time to complete:
+           --
+           -- - 'Async.Handle.async reads and writes'
+           -- - 'Async.Handle.PingPongPipelined test'
+           -- - 'Async.Socket.PingPong test'
+           -- - 'Async.Socket.PingPongPipelined test'
+           --
+           -- or not terminate at all within reasonable timeout:
+           --
+           -- - 'Async.Handle.PingPong test'
+           --
+           -- Note that in the test setup, we start io manager thread for each
+           -- test case.  If we don't terminate it when the iocp port is closed
+           -- we might end up with a busy loop which runs expensive system
+           -- calls that fail.  This could explain the slowdown.
+           --
+           -- In every observed case both values:
+           --
+           -- - 'gqcsResult'
+           -- - 'gqcsCompletionKey'
+           --
+           -- where false.
+           --
+               return ()
+         | gqcsOverlappedIsNull ->
+           -- 'GetLastCompletionStatus' has not dequeued any completion packet
+           -- from the completion port.  Must be the first clause, since if
+           -- this is not true we cannot trust other arguments.
+              throwIO NullOverlappedPointer
+         | not gqcsCompletionKey ->
+           -- skip if the completion key did not match with our
+           -- 'MAGIC_COMPLETION_KEY'
+            dequeueCompletionPackets iocp
+         | gqcsResult -> do
+           -- 'GetQueuedCompletionStatus' system call returned without errors.
+             !(numBytes :: Int) <-
+                 fromIntegral <$> peek numBytesPtr
+             mvarPtr <- peek (iodDataPtr gqcsIODataPtr)
+             mvar <- deRefStablePtr mvarPtr
+             freeStablePtr mvarPtr
+             hp <- Win32.getProcessHeap
+             Win32.heapFree hp 0 (castPtr gqcsIODataPtr)
```

Comment:
If we use local allocation in `withIOData` then we will not need to heap free here.

Reply:
But we cannot use local allocations, as discussed.